Well, on reviewing the code a bit more closely it seems that this part doesn't run until run time, so... I guess I don't know what's actually going on here.  Note that the failure message looks like this:

```
[ERROR] Failures: 
[ERROR]   PropertyInjectionTestCase.testConfigPropertyInjectedIntoTest:17 expected: <A message on 8081> but was: <A message>
[ERROR]   ServletTestCase.testServlet:33 1 expectation failed.
Response body doesn't match expectation.
Expected: is "A message"
  Actual: A message on null
```

That's two failures.  The first one looks like it didn't write the correct message at all - I mean "on" is missing so the output it expected is not the one that was produced by your change.  Then the second has the opposite problem.  Maybe the test is bad?

--- a/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ConfigBuildStep.java
+++ b/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ConfigBuildStep.java
@@ -23,10 +23,10 @@ import io.quarkus.arc.processor.BeanDeploymentValidator;
 import io.quarkus.arc.processor.DotNames;
 import io.quarkus.arc.processor.InjectionPointInfo;
 import io.quarkus.arc.runtime.ConfigDeploymentTemplate;
+import io.quarkus.arc.runtime.QuarkusConfigProducer;
 import io.quarkus.deployment.annotations.BuildProducer;
 import io.quarkus.deployment.annotations.BuildStep;
 import io.quarkus.deployment.annotations.Record;
-import io.smallrye.config.inject.ConfigProducer;
 
 /**
  * MicroProfile Config related build steps.
@@ -37,7 +37,7 @@ public class ConfigBuildStep {
 
     @BuildStep
     AdditionalBeanBuildItem bean() {
-        return new AdditionalBeanBuildItem(ConfigProducer.class);
+        return new AdditionalBeanBuildItem(QuarkusConfigProducer.class);
     }
 
     @BuildStep

--- a/extensions/arc/runtime/src/main/java/io/quarkus/arc/runtime/ConfigDeploymentTemplate.java
+++ b/extensions/arc/runtime/src/main/java/io/quarkus/arc/runtime/ConfigDeploymentTemplate.java
@@ -24,7 +24,7 @@ import java.util.Set;
 import javax.enterprise.inject.spi.DeploymentException;
 
 import org.eclipse.microprofile.config.Config;
-import org.eclipse.microprofile.config.ConfigProvider;
+import org.eclipse.microprofile.config.spi.ConfigProviderResolver;
 
 import io.quarkus.runtime.annotations.Template;
 
@@ -35,7 +35,7 @@ import io.quarkus.runtime.annotations.Template;
 public class ConfigDeploymentTemplate {
 
     public void validateConfigProperties(Map<String, Set<Class<?>>> properties) {
-        Config config = ConfigProvider.getConfig();
+        Config config = ConfigProviderResolver.instance().getConfig();
         for (Entry<String, Set<Class<?>>> entry : properties.entrySet()) {
             Set<Class<?>> propertyTypes = entry.getValue();
             for (Class<?> propertyType : propertyTypes) {

--- a/extensions/arc/runtime/src/main/java/io/quarkus/arc/runtime/QuarkusConfigProducer.java
+++ b/extensions/arc/runtime/src/main/java/io/quarkus/arc/runtime/QuarkusConfigProducer.java
@@ -1,0 +1,95 @@
+package io.quarkus.arc.runtime;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.context.Dependent;
+import javax.enterprise.inject.Produces;
+import javax.enterprise.inject.spi.InjectionPoint;
+
+import org.eclipse.microprofile.config.Config;
+import org.eclipse.microprofile.config.inject.ConfigProperty;
+import org.eclipse.microprofile.config.spi.ConfigProviderResolver;
+
+import io.smallrye.config.inject.ConfigProducerUtil;
+
+/**
+ * This class is the same as io.smallrye.config.inject.ConfigProducer
+ * but uses the proper Quarkus way of obtaining org.eclipse.microprofile.config.Config
+ */
+@ApplicationScoped
+public class QuarkusConfigProducer {
+
+    @Produces
+    Config getConfig(InjectionPoint injectionPoint) {
+        return ConfigProviderResolver.instance().getConfig();
+    }
+
+    @Dependent
+    @Produces
+    @ConfigProperty
+    String produceStringConfigProperty(InjectionPoint ip) {
+        return ConfigProducerUtil.getValue(ip, String.class, getConfig(ip));
+    }
+
+    @Dependent
+    @Produces
+    @ConfigProperty
+    Long getLongValue(InjectionPoint ip) {
+        return ConfigProducerUtil.getValue(ip, Long.class, getConfig(ip));
+    }
+
+    @Dependent
+    @Produces
+    @ConfigProperty
+    Integer getIntegerValue(InjectionPoint ip) {
+        return ConfigProducerUtil.getValue(ip, Integer.class, getConfig(ip));
+    }
+
+    @Dependent
+    @Produces
+    @ConfigProperty
+    Float produceFloatConfigProperty(InjectionPoint ip) {
+        return ConfigProducerUtil.getValue(ip, Float.class, getConfig(ip));
+    }
+
+    @Dependent
+    @Produces
+    @ConfigProperty
+    Double produceDoubleConfigProperty(InjectionPoint ip) {
+        return ConfigProducerUtil.getValue(ip, Double.class, getConfig(ip));
+    }
+
+    @Dependent
+    @Produces
+    @ConfigProperty
+    Boolean produceBooleanConfigProperty(InjectionPoint ip) {
+        return ConfigProducerUtil.getValue(ip, Boolean.class, getConfig(ip));
+    }
+
+    @Dependent
+    @Produces
+    @ConfigProperty
+    <T> Optional<T> produceOptionalConfigValue(InjectionPoint injectionPoint) {
+        return ConfigProducerUtil.optionalConfigValue(injectionPoint, getConfig(injectionPoint));
+    }
+
+    @Dependent
+    @Produces
+    @ConfigProperty
+    <T> Set<T> producesSetConfigPropery(InjectionPoint ip) {
+        return ConfigProducerUtil.collectionConfigProperty(ip, getConfig(ip), new HashSet<>());
+    }
+
+    @Dependent
+    @Produces
+    @ConfigProperty
+    <T> List<T> producesListConfigPropery(InjectionPoint ip) {
+        return ConfigProducerUtil.collectionConfigProperty(ip, getConfig(ip), new ArrayList<T>());
+    }
+
+}

--- a/integration-tests/main/src/main/java/io/quarkus/it/rest/ServiceWithConfig.java
+++ b/integration-tests/main/src/main/java/io/quarkus/it/rest/ServiceWithConfig.java
@@ -1,0 +1,23 @@
+package io.quarkus.it.rest;
+
+import javax.enterprise.context.ApplicationScoped;
+
+import org.eclipse.microprofile.config.inject.ConfigProperty;
+
+@ApplicationScoped
+public class ServiceWithConfig {
+
+    @ConfigProperty(name = "quarkus.http.host")
+    String quarkusHost;
+
+    @ConfigProperty(name = "web-message")
+    String message;
+
+    public String host() {
+        return quarkusHost;
+    }
+
+    public String message() {
+        return message;
+    }
+}

--- a/integration-tests/main/src/main/java/io/quarkus/it/rest/TestResource.java
+++ b/integration-tests/main/src/main/java/io/quarkus/it/rest/TestResource.java
@@ -64,6 +64,9 @@ public class TestResource {
     @Inject
     ExternalService service;
 
+    @Inject
+    ServiceWithConfig config;
+
     private final AtomicInteger count = new AtomicInteger(0);
 
     @GET
@@ -75,6 +78,18 @@ public class TestResource {
     @Path("/service")
     public String service() {
         return service.service();
+    }
+
+    @GET
+    @Path("/config/host")
+    public String configPort() {
+        return config.host();
+    }
+
+    @GET
+    @Path("/config/message")
+    public String configMessage() {
+        return config.message();
     }
 
     @GET

--- a/integration-tests/main/src/test/java/io/quarkus/it/main/JaxRSTestCase.java
+++ b/integration-tests/main/src/test/java/io/quarkus/it/main/JaxRSTestCase.java
@@ -42,6 +42,16 @@ public class JaxRSTestCase {
     }
 
     @Test
+    public void testConfigInjectionOfPort() {
+        RestAssured.when().get("/test/config/host").then().body(is("0.0.0.0"));
+    }
+
+    @Test
+    public void testConfigInjectionOfMessage() {
+        RestAssured.when().get("/test/config/message").then().body(is("A message"));
+    }
+
+    @Test
     public void testAnnotatedInterface() {
         RestAssured.when().get("/interface").then().body(is("interface endpoint"));
 

